### ISSUE TITLE
[node] `/tx` endpoint returns `-1` for the `fee`

### ISSUE THREAD

**Issue by osagga** [bitpay/bitcore#2245]
I'm using the bitcore node as of this commit 0b5b4210168b622edac2bbbfa039bec9e2aa098f (latest as of now) on `regtest`, my setup is as follows:

```
$node -v
v10.16.0

$ ./bitcoind --version
Bitcoin Core Daemon version v0.18.0
```
When I try to query for a transaction, I get `-1` for the `fee` value, this value used to be correct earlier (at least as early as commit 7a4bc3bbb12a08c3297398a5b0ee1e7d563d5a2e)

Here's an example where fee is `-1` using the latest bitcore node
```json
curl localhost:3000/api/BTC/regtest/tx/d7e7fae59c4dcfe7b14a64c756d66ea83fb475fe5de80f955ff6f909297090f8/ | python -m json.tool
{
    "_id": "5d0a986998032f7e182e8186",
    "blockHash": "650c6e8da4a64fb479b9d1cad057ff78e26013672d076cf1f2be0fd959841e0f",
    "blockHeight": 626,
    "blockTime": "2019-06-19T20:18:29.000Z",
    "blockTimeNormalized": "2019-06-19T20:18:29.004Z",
    "chain": "BTC",
    "coinbase": false,
    "confirmations": 7,
    "fee": -1,
    "inputCount": 161,
    "locktime": 625,
    "network": "regtest",
    "outputCount": 2,
    "size": 10380,
    "txid": "d7e7fae59c4dcfe7b14a64c756d66ea83fb475fe5de80f955ff6f909297090f8",
    "value": 300312104750
}
```

Here's an example when the `fee` value used to work, note that I'm using bitcoind version `0.17.1` and node version `8.94` here
```json
curl -q localhost:3000/api/BTC/regtest/tx/5c0d2a8f73c543cdfac8d93f354c81055003339b08cdb4ccba6c9ea3a665e2b7/ | python -m json.tool
{
    "_id": "5d0aa3c780d1f92609e04ca8",
    "blockHash": "3d32810cb4909dcf7f45838fca00f329d94fd58a4d1d9d4ebcddff005d334fdc",
    "blockHeight": 315,
    "blockTime": "2019-06-19T21:06:53.000Z",
    "blockTimeNormalized": "2019-06-19T21:06:53.001Z",
    "chain": "BTC",
    "coinbase": false,
    "confirmations": 6,
    "fee": 4480,
    "inputCount": 1,
    "locktime": -1,
    "network": "regtest",
    "outputCount": 2,
    "size": 223,
    "txid": "5c0d2a8f73c543cdfac8d93f354c81055003339b08cdb4ccba6c9ea3a665e2b7",
    "value": 333749943910
}
```

 What could be causing this? Is this a bug in `bitcore-node`? Or caused by updating bitcoind (a broken RPC or something)?

**Comment by micahriggan** [comment]
Was this a zero fee transaction? If not, then there maybe an issue in models/transaction.ts related to calculating the fee.

**Comment by osagga** [comment]
@micahriggan 
I dont think its a zero fee transaction, here's another example (on `regtest`):

```json
sh-4.2$ curl -s localhost:3000/api/BTC/regtest/tx/8dd753b13f8112fc7b9a377be4f95c292b6e8c49768e85a940c41b7e00764941 | python -m json.tool
{
    "_id": "5d0bb2edd1058c9fb2118cbe",
    "blockHash": "3cef934aae0d63d5a8693680994c2a2464dc15dc99bcb277881638de4cd90aa3",
    "blockHeight": 313,
    "blockTime": "2019-06-20T16:23:47.000Z",
    "blockTimeNormalized": "2019-06-20T16:23:47.005Z",
    "chain": "BTC",
    "coinbase": false,
    "confirmations": 6,
    "fee": -1,
    "inputCount": 1,
    "locktime": -1,
    "network": "regtest",
    "outputCount": 2,
    "size": 223,
    "txid": "8dd753b13f8112fc7b9a377be4f95c292b6e8c49768e85a940c41b7e00764941",
    "value": 332812453910
}
sh-4.2$ curl -s localhost:3000/api/BTC/regtest/tx/8dd753b13f8112fc7b9a377be4f95c292b6e8c49768e85a940c41b7e00764941/coins | python -m json.tool
{
    "inputs": [
        {
            "_id": "5d0bb2e7d1058c9fb2118205",
            "address": "mtgDP8CdcG4KdQMs6g68rvyt18aT468Yfq",
            "chain": "BTC",
            "coinbase": false,
            "confirmations": -1,
            "mintHeight": 307,
            "mintIndex": 0,
            "mintTxid": "16fc5e4b0d3fe52bb6b59d41d13afab2b9dd9f2c890962e0e9fdf75c5f0e8c8f",
            "network": "regtest",
            "script": "76a9149059f92379d7e43ac9c050eee42fa1a5bfe5001888ac",
            "spentHeight": 313,
            "spentTxid": "8dd753b13f8112fc7b9a377be4f95c292b6e8c49768e85a940c41b7e00764941",
            "value": 332812458390
        }
    ],
    "outputs": [
        {
            "_id": "5d0bb2edd1058c9fb2118cb8",
            "address": "mtgDP8CdcG4KdQMs6g68rvyt18aT468Yfq",
            "chain": "BTC",
            "coinbase": false,
            "confirmations": -1,
            "mintHeight": 313,
            "mintIndex": 0,
            "mintTxid": "8dd753b13f8112fc7b9a377be4f95c292b6e8c49768e85a940c41b7e00764941",
            "network": "regtest",
            "script": "76a9149059f92379d7e43ac9c050eee42fa1a5bfe5001888ac",
            "spentHeight": -2,
            "spentTxid": "",
            "value": 332810557895
        },
        {
            "_id": "5d0bb2edd1058c9fb2118cb9",
            "address": "2MskwaoyP1qiAJDvUp8yXqvZFwPLn1bFQNF",
            "chain": "BTC",
            "coinbase": false,
            "confirmations": -1,
            "mintHeight": 313,
            "mintIndex": 1,
            "mintTxid": "8dd753b13f8112fc7b9a377be4f95c292b6e8c49768e85a940c41b7e00764941",
            "network": "regtest",
            "script": "a914059ec8ee8e19d9c68ff609ff101b810e91fd761087",
            "spentHeight": 316,
            "spentTxid": "3f480edc7bc340fb2dfbf77b715dabed0114434b3a8394c99ce52eb8ceaee25e",
            "value": 1896015
        }
    ]
}
```

If you subtract the input from the output you get
```
332812458390 - (332810557895 + 1896015) = 4480
```
So the fee should be 4480 sats.

Note that older versions of v8 don't have this issue (as of the commit mentioned above), so I doubt that this is related to the way the fee is calculated
https://github.com/bitpay/bitcore/blob/9ccfa92fe502774ae3d280a1d217c92822b492fa/packages/bitcore-node/src/models/transaction.ts#L289-L296
This logic hasn't been modified over the past 6 months, so the problem has to be coming from somewhere else, maybe the above logic is never triggered (all the transactions have `-1` fees for me)?

**Comment by micahriggan** [comment]
Turns out fee calculation was broken in c3b78769a13fea6f8126b58cce1eba6f994bb48e.

Changing to splice forced a re-order of addTransactions to the top, to retain access to params.mintOpts,

Unfortunately addTransactions requires that the spends have been processed, so that we can find all coins spent in this block, to calculate the tx fee.

**Comment by osagga** [comment]
@micahriggan oh that's why! I was debugging and I noticed that the fees are -1 because `groupedSpends` is `undefined`. It makes sense why now, thanks for the update, hopefully #2247 gets merged in soon. 
Once its merged, I'll test and confirm the fix (and close the issue)

**Comment by osagga** [comment]
@micahriggan is there an ETA on when would the fee fix be applied? the current master branch doesn't have the fix, I would love for it to be fixed, and maybe make a new release?

**Comment by osagga** [comment]
Fixed by #2247, thanks!

#close
